aten/src/ATen/core/TensorAccessor.h

```diff
@@ -91,56 +94,80 @@ class TensorAccessor<T,1,PtrTraits> : public TensorAccessorBase<T,1,PtrTraits> {
 // Use RestrictPtrTraits as PtrTraits if you want the tensor's data pointer to be marked as __restrict__.
 // Instantiation from data, sizes, strides is only needed on the host and std::copy isn't available
 // on the device, so those functions are host only.
-template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits>
+template<typename T, size_t N, template <typename U> class PtrTraits = DefaultPtrTraits, typename index_t = int64_t>
 class PackedTensorAccessorBase {
 public:
   typedef typename PtrTraits<T>::PtrType PtrType;
-  AT_HOST PackedTensorAccessorBase(PtrType data_, const int64_t * sizes_, const   int64_t * strides_)
+  AT_HOST PackedTensorAccessorBase(PtrType data_, const index_t * sizes_, const index_t * strides_)
   : data_(data_)
   {
     std::copy(sizes_, sizes_ + N, std::begin(this->sizes_));
     std::copy(strides_, strides_ + N, std::begin(this->strides_));
   }
-  AT_HOST_DEVICE int64_t stride(int64_t i) const { return strides_[i]; }
-  AT_HOST_DEVICE int64_t size(int64_t i) const { return sizes_[i]; }
+
+  // if index_t is not int64_t, we want to have an int64_t constructor
+  template <typename source_index_t, class = typename std::enable_if<std::is_same<source_index_t, int64_t>::value>::type>
```

Comment:
I think this would be clearer if you get rid of the `typename source_index_t` arg, and only enable if `index_t != int64_t`.

Reply:
I would need a hint how to do this. I tried all sorts of things and it didn't work out.
(But looking at the methods, I should probably use index_t more...)

Reply:
Does this not work?
```
template<typename = std::enable_if<!std::is_same<index_t, int64_t>::value>::type>
```

Reply:
No. I did try that. :) It gives:
`TensorAccessor.h(109): error: class "std::enable_if<false, void>" has no member "type"`
As far as I understand it, the reason is that you can have an "impossible template" (which is what enable_if produces in the false case) only as long as the template parameters are not fully given. So I need some unspecified bit, and after trying out various variants and trying to take something that looks similar to what `rgrep enable_if` gives, I settled on this. I don't like it, but it seems to not be too harmful.

Reply:
I see. I've found that this works, but not sure if it is less ugly:
```
template<typename index_tt = index_t>
std::enable_if<!std::is_same<index_tt, int64_t>::value>::type
```

See https://godbolt.org/z/bIX4_O .

Keep it as-is also SGTM. BTW, if you want to use `enable_if_t`, we have it as `c10::guts::enable_if_t`.

Reply:
I think that that introduces a return type for `f` which isn't applicable for the constructor.